### ISSUE TITLE
Discriminated union types not resolved properly with ngSwitch

### ISSUE THREAD

**Issue by jinbijin** [angular/angular#34522]
# 🐞 bug report

### Affected Package
```
@angular/compiler
@angular/language-service
```

### Is this a regression?
No

### Description
In templates, type resolution fails when using `ngSwitch` on the discriminant property of a discriminated union type; in all cases, the type seems to be resolved to the first option.

## 🔬 Minimal Reproduction
The following example uses an enum value as the discriminant, but the same behaviour is seen when using string values as the discriminant.
See https://github.com/jinbijin/minimal-example for the full repository; it's a new Angular project except for:
`app.component.ts`:
```
import { Component } from '@angular/core';

enum MaybeType {
  Nothing = 0,
  Just = 1
}

interface Nothing {
  type: MaybeType.Nothing;
}

interface Just<T> {
  type: MaybeType.Just;
  value: T;
}

type Maybe<T> = Nothing | Just<T>;

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent {
  MaybeType = MaybeType;
  maybeText: Maybe<string> = { type: MaybeType.Just, value: 'hello!' };
}
```

`app.component.html`:
```
<ng-container [ngSwitch]="maybeText.type">
  <span *ngSwitchCase="MaybeType.Nothing">Nothing</span>
  <span *ngSwitchCase="MaybeType.Just">{{ maybeText.value }}</span>
</ng-container>

<router-outlet></router-outlet>
```

Expected behaviour: Project compiles without any errors. With strict full template type checking the language service doesn't generate errors.

Actual behaviour: The following error is generated on running `ng serve` with full template type checking:
```
ERROR in src/app/app.component.html:3:43 - error TS-992339: Property 'value' does not exist on type 'Maybe<string>'.
  Property 'value' does not exist on type 'Nothing'.

3   <span *ngSwitchCase="MaybeType.Just">{{ maybeText.value }}</span>
                                            ~~~~~~~~~~~~~~~~

  src/app/app.component.ts:21:16
    21   templateUrl: './app.component.html',
                      ~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component AppComponent.
```
With strict template checking, the language service generates the following error:
```
Identifier 'value' is not defined. 'Maybe<string>' does not contain such a member
```

## 🌍  Your Environment

**Angular Version:**
```
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 9.0.0-rc.7
Node: 12.6.0
OS: darwin x64

Angular: 9.0.0-rc.7
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-rc.7
@angular-devkit/build-angular     0.900.0-rc.7
@angular-devkit/build-optimizer   0.900.0-rc.7
@angular-devkit/build-webpack     0.900.0-rc.7
@angular-devkit/core              9.0.0-rc.7
@angular-devkit/schematics        9.0.0-rc.7
@ngtools/webpack                  9.0.0-rc.7
@schematics/angular               9.0.0-rc.7
@schematics/update                0.900.0-rc.7
rxjs                              6.5.3
typescript                        3.6.4
webpack                           4.41.2
```

**Anything else relevant?**
tsconfig.json:
```
{
  "compileOnSave": false,
  "compilerOptions": {
    "baseUrl": "./",
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "downlevelIteration": true,
    "experimentalDecorators": true,
    "module": "esnext",
    "moduleResolution": "node",
    "importHelpers": true,
    "target": "es2015",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2018",
      "dom"
    ]
  },
  "angularCompilerOptions": {
    "fullTemplateTypeCheck": true,
    "strictInjectionParameters": true,
    "strictTemplates": true
  }
}
```

**Comment by alexzuza** [comment]
Minimal reproduction https://ng-run.com/edit/8ZCWt14EFPjErwI54zah?open=app%2Fapp.component.html

**Comment by Dassderdie** [comment]
Is this a duplicate of https://github.com/angular/angular/issues/20780 ?

**Comment by jinbijin** [comment]
Yup, looks like it's the same issue, but in this case it affects the compiler as well

**Comment by alxhub** [comment]
I wonder if this is something we can fix in the type definitions for `NgSwitch`, with the tools we have today. Or if this is going to require some special casing.

**Comment by zlepper** [comment]
This is something we would really love to see fixed. We use discriminated unions very heavily in our code, which also means we can't turn on the strict type checking yet, even though it would provide a tremendous value for us.

**Comment by grantwhiting** [comment]
Has there been any movement on this recently? Thanks!

**Comment by JoostK** [comment]
@grantwhiting Nope, this one hasn't been figured out yet. I don't see this being modeled using the currently available type checking tools, as `ngSwitchCase` expressions have to be combined with the corresponding `ngSwitch` case during type checking codegen, so I think there is a need for some custom codegen logic.

**Comment by Maximaximum** [comment]
I've created a stackblitz to report this issue, and only after that I was able to find this issue that's already open. Anyway, just in case it helps, here's my stackblitz: https://stackblitz.com/edit/angular-issue-ngswitch-type-narrowing

This issue is quite annoying, so I really hope that it gets enough attention from the Angular team and gets fixed soon.

**Comment by GabrielGil** [comment]
This is an actually really annoying issue which happens with any GraphQL generated schema. It also prevents my team from using strict template checking.

We hope this is addressed soon.

Our workaround for the time being (instead of disabling template checking completely) is to set the type of these values to `any`. Is there any better solution to apply until this is resolved?

**Comment by alxhub** [comment]
This is a big challenge, mainly as it's a design limitation with how `NgSwitch` works. I think we should look into ways of doing this, but realistically I don't think this is something we can address quickly.

**Comment by yankee42** [comment]
> This is a big challenge [..] but realistically I don't think this is something we can address quickly.

Maybe some way to cast values in templates could be a workaround?
```html
<span *ngSwitchCase="MaybeType.Just" ngCast="maybeText as Just">{{ maybeText.value }}</span>
```

Of course right now we can use [$any](https://angular.io/guide/template-typecheck#disabling-type-checking-using-any):
```html
<span *ngSwitchCase="MaybeType.Just">{{ $any(maybeText).value }}</span>
```
As long as it is as simple as this example this works, although it completely looses type-checks. However when the element contains many references to the type in question, the template becomes quickly littered with `$any`-calls. Some way to cast for a whole element would reduce this.

**Comment by Airblader** [comment]
> Maybe some way to cast values in templates could be a workaround?

Introducing workarounds as features is not a good idea either. There's also other issues, however, that pitch the idea of bringing types into the templates in some way.

In any case, type-casting like that is already possible in a slightly more verbose way (consider replacing ngIf with something that can deal with falsy values). Just be aware that at the end of the day this is still just a lie to the compiler and not bullet-proof.

```
public readonly asFoo = (x: unknown): Foo => x as Foo;
```

```
<ng-container *ngIf="asFoo(untypedX) as x">
  {{ x.notInFoo }} <!-- error -->
</ng-container>
```

Personally I use this "trick" everytime I need to type the context variables of an ng-template which would otherwise just be `any`.

**Comment by amakhrov** [comment]
> at the end of the day this is still just a lie to the compiler 

We can move the `if` condition into the asFoo method, restoring type safety (note the absence of type assertion):
```
public readonly asFoo(x: Foo | Bar): Foo undefined => x.type === 'foo' ? x : undefined;
```
```
<ng-container *ngIf="asFoo(fooOrBar) as x">
  {{ x.inFoo }}
</ng-container>
```

A drawback, as in the previous comment, is that it "swallows" falsy values.

Furthermore, it's possible to make this `asFoo` function generic (still preserving type safety), so that we can define it in the component once, and use for all possible alternatives:
```
<ng-container *ngIf="asType('foo', fooBarBaz) as foo">
  {{ foo.inFoo }}
</ng-container>
<ng-container *ngIf="asType('bar', fooBarBaz) as bar
  {{ foo.inBar }}
</ng-container>
```

**Comment by Airblader** [comment]
This doesn't really solve type safety for the general ng-template case I mentioned. The context variables (let) are typed any and that will always type check when passed to the asFoo function, regardless of whether it's correct. But that issue is more general and not specific to ngSwitch

**Comment by simeyla** [comment]
@amakhrov can you provide your implementation for `asType`. Also have you considered situations where the type is a discriminated union without distinct named types for each sub-type.

eg. 

   type Animal = { type: 'dog', sound: 'bark' | 'growl' } | { type: 'cat', sound: 'purr' | 'meow' };

I tried doing this:

```
     public asType<T extends Animal['type']>(type: T, x: Animal) : (Animal & { type: T })
    {
        if (!x)
        {
            return x;
        }
        else if (x.type === type) 
        {
            return <any>x;
        }
        else 
        {
            throw new Error('Not expected type');
        }
    }  
```

I throw an error if the type is not as expected (for a switch statement I think that's what you would want), and I constrain the response to the correct type in the return signature (`Animal & { type: T }`).

This works fine in typescript code, but unfortunately not inside a template.

Was your solution something like this? I feel like I'm either close, or trying to do something impossible.

**Comment by amakhrov** [comment]
> (Animal & { type: T })

This is not what we want to represent. instead, we want to have a subtype of Animal that satisfies the constraint `type: 'some-type'`.

You can see example implementation [in the playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDeAoKVSQFxQOQ1wBo0oBDHAZ2ACcBLAOwHMUBfFFTaAIQWRM5y4uREgCNKNBszYdw0AMK9U6AXnkj0AYwl0mrdp1hdFiOAB8oPC-INyoAZQCuozgB4AqvVoB7eoSgAIjQA0hAgUBAAHsAQ9AAmFFAA1mHeAGZQnj5+gTQAaqQANo7QUTHxiVm+ANpB1KEgALoAfLwkVfQR0bEJUAAUDGkQ1FAAkvRD1NQQcQCUUCToAPxjE8PTcV3lvR0L6PtQK+OTG7UhYY1bPYl1BcXQiweHqyczjwc49BAAbsPvUJ8fn8PlAvr9qABuWyQBzOTgAeTSMC4rgAKlcKkZ5NVcJxcC1eE4XHJXMj5P5cXIiFBUc1odB7AALbyOQpxLgQOCIIkIpEo-C4OmGJkstkcnjcuFyRHI1xCQXsTS+KhkCg8uQAMVImmA3mo4UQaUc9B12Q8Xl8zT682UUGmwEc1E6rjqDQxvRSIHSmQt9CtnpwrrC83grUeaPdlV9Z3qFytcRoOHRFlR1Ua-mN2RwHVmOHVkHN2X8QZA-lpUAsxriEDSDBmCFatoOSvoKoT1BgU14neopBAADpaBQe32+u35it2wCoNV240oU8oLQMmOaD3B-Eooi+pmap7LqREqj5gBCeCIAC0AEYbf90AB6e9QMDecrAWhFQrhY0UUhDDB2IeFDDO+vhQBeUCiI4wBQAA7tAST0N4sFwYypAwfBuDTFAcTeFId52hADpOlAu6dIesLEgWHTFucpY0s0C5PGwi72o6nRVjWdZxEx6BsDIhREaRvonjgZJQigLYqoe+bQIgMlSpAWo6nqICksYVqzH0lKQLgsyScqMGGIgAoGa2MGkLwClURAfScBmvqzBCUCPrAFbCdWtZfHEZnSZoVlqoptnVAARDAIX+CF8ghemwnZE5LlPuYUCKJWm5eW8QA)

I also created  gist with usage example https://gist.github.com/amakhrov/a7b2a30b13668f714d40e095db437486

Hope this helps.

**Comment by striky1** [comment]
@amakhrov 

> I also created gist with usage example https://gist.github.com/amakhrov/a7b2a30b13668f714d40e095db437486

I tried your solution and it works fine for compiler but IDE (PHPStorm/WebStorm) does not recognize which types I'm trying to have.

```
type TypeA = {vmType: 'type-A', propertyA: string};
type TypeB = {vmType: 'type-B', propertyB: string};

types: (TypeA | TypeB) = {vmType: 'type-A', propertyA: ''};
```

```
<ng-container *ngIf="asSubtype('type-A', types) as newType">
    {{newType.propertyB}}  <= IDE don't care about it. Property newType is TypeA or TypeB type. Compiler throw error
</ng-container>
```

Anyway, same result is with:

```
static instanceOfTypeA(types: any): types is TypeA {
  return types.vmType === 'type-A';
}
```

```
<ng-container *ngIf="instanceOfTypeA(types)">
    {{types.propertyB}}  <= IDE don't care about it. Property types is TypeA or TypeB type. Compiler throw error
</ng-container>
```

Originally I was wonder how to resolve it 'cause I have multiple types in Array and seems that your or `object is type` function are as possible temporary solution. Disadvantage is that we can't use them with `*ngSwitch` what means that on each DC Angular must check all `*ngIf` conditions (based on Array changes).

**Comment by amakhrov** [comment]
> works fine for compiler but IDE (PHPStorm/WebStorm) does not recognize

Yep, it's the same for me. WebStorm still thinks that `newType` variable is of type `TypeA | TypeB` - even though everything compiles properly. Not sure if it's an issue with angular language server or with webstorm itself though.

> Anyway, same result is with:
> *ngIf="instanceOfTypeA(types)"

exactly, because ngIf doesn't do type narrowing. otherwise there would be no problem at all to discuss here!

**Comment by Harpush** [comment]
This is really important... The amount of if else in template based on discriminate union that needs those hacks is too much to not fix it...

**Comment by jaufgang** [comment]
The solution to this problem is to use `*ngIf` with a pipe that evaluates a type guard and returns the original object with a narrowed type if the object passes the type guard and undefined if it doesn't 

```js
// guard-type.pipe.ts

export type TypeGuard<A, B extends A> = (a: A) => a is B;

@Pipe({
  name: 'guardType'
})
export class GuardTypePipe implements PipeTransform {

 transform<A, B extends A>(value: A, typeGuard: TypeGuard<A, B>): B | undefined {
    return typeGuard(value) ? value : undefined;
  }

}
```
This pipe effectively narrows the type without "lying to the compiler". (it actually performs proper run-time type checking via the type guard function).  It will also throw an error if the object being type-checked cannot be narrowed by the type guard (i.e. it is not a member of the union type being narrowed)


Sample usage: 
```js
// shapes.type.ts

import { TypeGuard } from "./guard-type.pipe";

export interface Square {
  kind: "square";
  size: number;
}

export interface Rectangle {
  kind: "rectangle";
  width: number;
  height: number;
}

export interface Circle {
  kind: "circle";
  radius: number;
}

export type Shape = Square | Rectangle | Circle;

export const isSquare: TypeGuard<Shape, Square> = (
  shape: Shape
): shape is Square => shape.kind === "square";

export const isRectangle: TypeGuard<Shape, Rectangle> = (
  shape: Rectangle
): shape is Rectangle => shape.kind === "rectangle";

export const isCircle: TypeGuard<Shape, Circle> = (
  shape: Shape
): shape is Circle => shape.kind === "circle";
```

```js
// app.component.ts

import { Component, VERSION } from "@angular/core";
import { isCircle, isRectangle, isSquare, Shape } from "./shapes.type";

@Component({
  selector: "my-app",
  templateUrl: "./app.component.html",
  styleUrls: ["./app.component.css"]
})
export class AppComponent {
  shapes: Shape[] = [
    {
      kind: "square",
      size: 5
    },
    {
      kind: "circle",
      radius: 10
    },
    {
      kind: "rectangle",
      height: 7,
      width: 9
    },
    {
      kind: "rectangle",
      height: 8.5,
      width: 11
    },
    {
      kind: "circle",
      radius: 8
    }
  ];

  // set the imported type guard functions as class properties
  // so they can be passed by the template to the guardType pipe
  isCircle = isCircle;
  isRectangle = isRectangle;
  isSquare = isSquare;
}

```

```html
<!-- app.component.html -->
<ul>
  <li *ngFor="let shape of shapes">
    <strong>{{shape.kind}}</strong>:

    <span *ngIf="shape| guardType: isCircle as circle">
      <!-- circle is strongly typed as a Circle -->
      radius = {{circle.radius}}
    </span>

    <span *ngIf="shape| guardType: isSquare as square">
      <!-- square is strongly typed as a Square -->
      size = {{square.size}}
    </span>

    <span *ngIf="shape| guardType: isRectangle as rectangle">
      <!-- rectangle is strongly typed as a Rectangle -->
      width = {{rectangle.width}}, height = {{rectangle.height}}
    </span>
  </li>
</ul>
```
which renders as:
![image](https://user-images.githubusercontent.com/1747254/105063628-ef5e9e00-5a49-11eb-961e-eb0778fad5c1.png)

I created a StackBlitz to demonstrate this live in action.  SeeL https://stackblitz.com/edit/angular-guard-type-pipe

**Comment by Lonli-Lokli** [comment]
> This doesn't really solve type safety for the general ng-template case I mentioned. The context variables (let) are typed any and that will always type check when passed to the asFoo function, regardless of whether it's correct. But that issue is more general and not specific to ngSwitch

@Airblader is there an issue for typed let variables ? We use template guards but it does not help in all the cases

**Comment by Airblader** [comment]
@Lonli-Lokli Yeah, there is https://github.com/angular/angular/issues/28731

**Comment by alechemy** [comment]
@jaufgang I downloaded your stackblitz and opened it in VS Code, and ran into some issues. It doesn't appear that the type is being correctly inferred when using the pipe:
<img width="676" alt="Screen Shot 2021-03-18 at 15 31 57@2x" src="https://user-images.githubusercontent.com/22478984/111687328-881e5980-8800-11eb-9ed1-fad070256354.png">

Meanwhile, if I simply call the guard directly, it works OK: 
<img width="767" alt="Screen Shot 2021-03-18 at 15 32 20@2x" src="https://user-images.githubusercontent.com/22478984/111687292-7f2d8800-8800-11eb-992a-09520bf05a6b.png">

This carries over to builds as well: `ng build --configuration=prod` will incorrectly pass when using that pipe and accessing a non-existent member of the interface (e.g. `square.radius`), but will properly fail when calling the guard directly.

**Comment by jaufgang** [comment]
@alechemy, thank you for catching this bug.  It was an error in my sample type guard for the `Rectangle` type, not in the pipe itself.  

I had 
```js
export const isRectangle: TypeGuard<Shape, Rectangle> = (
  shape: Rectangle
): shape is Rectangle => shape.kind === "rectangle";
```
where it should have been 
```js
export const isRectangle: TypeGuard<Shape, Rectangle> = (
  shape: Shape
): shape is Rectangle => shape.kind === "rectangle";
```
(i.e. I had the input parameter as type `Rectangle` but it should have been `Shape` like in the other two type guards)

I've updated the stackBlitz and my above comment.  

Calling the type guard directly instead of through a pipe is not a good idea.  See the following article for a good explanation why: https://medium.com/showpad-engineering/why-you-should-never-use-function-calls-in-angular-template-expressions-e1a50f9c0496

**Comment by alechemy** [comment]
@jaufgang, yep, that fixed it. Thanks for looking into it. 

I am familiar with the advisory against invoking methods or getters directly in templates, though I am somewhat skeptical that it translates to poor performance here, given that the type guard method is so lightweight. I did a quick performance audit in my browser and didn't observe a meaningful difference between the two approaches (your 'pure pipe' vs directly calling the guard). 

In any case, thanks for putting the work into this workaround. Hopefully the underlying issue gets an actual fix soon.

FYI , running `npm install -S @angular/compiler@11.0.9` in the downloaded stackblitz was enough to get it to build cleanly for me.

**Comment by lukket** [comment]
@jaufgang thanks for this easy and reusable solution. I adapted it a little bit so it also works for unknown data as well.

```ts
export type TypeGuard<T> = (a: unknown) => a is T;

@Pipe({
  name: 'guardType',
})
export class GuardTypePipe implements PipeTransform {
  transform<T>(value: unknown, typeGuard: TypeGuard<T>): T | undefined {
    return typeGuard(value) ? value : undefined;
  }
}
```



https://angular-guard-type-pipe-ahwgyp.stackblitz.io

**Comment by Maximaximum** [comment]
Any chances this annoying bug is going to be fixed any time soon?

**Comment by spierala** [comment]
In some cases your ngSwitch can easily be replaced with ngIfs...

Calling Typescript [_type predicates_](https://languageimperfect.com/2019/08/05/understanding-typescript-type-predicates.html) from the template will help to get the correct type in the template.

Also see this stackblitz: https://stackoverflow.com/a/47221093

The function calls are usually simple, so there should not be any performance impacts.

```ts
export class MyComponent {
  isSubThing(t: Thing): t is SubThing {
    return t instanceof SubThing;
  }
}
```
Template:
```html
<div *ngIf="isSubThing(thing) && thing?.bar > 10">
  Conditional content...
</div>
```

**Comment by kemsky** [comment]
After experiments with `GuardTypePipe` I think that there is no any decent type inference in templates at all. ngIf, ngFor, ngSwitch - all break on any slightly generic or non-trivial code. Makes no sense to use pipe - it just does not do anything useful, just use functions:

```typescript
public readonly isCircle = (shape: Shape): Circle | undefined =>  shape.type === 'cirlce' ? shape : undefined;
```

```html
<ng-container [ngSwitch]="true">
      <circle *ngSwitchCase="isCircle(shape) as circle" [circle]="circle"/>
      <div *ngSwitchDefault>unknown</div>
</ng-container>
```

**Comment by amakhrov** [comment]
Looks like it will ultimately be solved with the new control flow syntax: https://github.com/angular/angular/discussions/50719

> The new syntax for template control flow has several major goals:
> ...
> 3. Works across multiple templates (if-elseif-else, switch-case-default), including template type-checking.

**Comment by JeanMeche** [comment]
Closing as this is now a solved problem with the new control flow syntax.
